src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+        /// Multiple comparisons are done by SortedSet to find if this Sample is present in the set.
+        /// </summary>
+        private readonly QubitTimeNode Sample = new QubitTimeNode(); // We reuse same object to avoid allocations.
+
+        public SortedQubitPool() {
+            NodeComparer = new VisitingComparer();
+            QubitsSortedByTime = new SortedSet<QubitTimeNode>(NodeComparer);
+        }
+
+        /// <summary>
+        /// Add Qubit to the set with specified time. Multiple qubits with the same time can be added.
+        /// Complexity is log(N), where N - number of nodes.
+        /// </summary>
+        /// <param name="qubitId">Id of a qubit to add to the set.</param>
+        /// <param name="time">Time of the qubit.</param>
+        public void Add(long qubitId, ComplexTime time) {
+            QubitTimeNode newNode = new QubitTimeNode(qubitId, time);
+            if (QubitsSortedByTime.Add(newNode)) {
+                // New item was added to the set, we are done.
+                return;
+            }
+            if (!QubitsSortedByTime.TryGetValue(newNode, out QubitTimeNode existingNode)) {
+                // We cannot add, but we cannot find a node with the same value. Is this a floating point glitch?
+                Debug.Assert(false, "Cannot add a value to SortedSet<QubitTimeNode> that isn't in the set.");
+                return;
+            }
+            // Add new node to the linked list as a second element.
+            newNode.NextNode = existingNode.NextNode;
+            existingNode.NextNode = newNode;
+        }
+
+        /// <summary>
+        /// Find qubit with the specified time in the set. If such qubit is not present,
+        /// find either maximum value less than it or minimum value greater than it.
+        /// This function uses one call to SortedSet.TryGetValue() so it takes log(N) time, where N - number of nodes.
+        /// </summary>
+        /// <param name="requestedTime">Sample time to find</param>
+        /// <param name="getLowerBound">"true" to find maximum value less than sample</param>
+        /// <param name="actualTime">Time found in the set</param>
+        /// <returns>"true" if sample was found in the set.</returns>
+        public bool FindBound(ComplexTime requestedTime, bool getLowerBound, out ComplexTime actualTime) {
+            // We use the following approach:
+            // We call function TryGetValue on a sorted set. If value is found, we go with it.
+            // Otherwise TryGetValue must make conclusion that the value is absent.
+            // Any comparison-based algorithm without caching must inspect both
+            // maximum value less than the sample and minimum value greater than the sample.
+            // So they must be among the values seen since reset.
+            // So we just need to harvest them from VisitingComparer.
+            Sample.Time = requestedTime;
+            NodeComparer.ResetForComparison(Sample);
```

Comment:
Oh! this is the line that confused me before: the value of Sample isn't used by NodeComparer and the reference to this.Sample isn't changing between invocations of FindBound... Would it make sense for the comparer to be configured with pointer(reference) to Sample during creation and ResetForComparison(Sample) changed to simply Reset()?

Reply:
This is a very good point! I actually considered both approaches and I believe the one I settled on is better. If you do not pass Sample to reset, there's a subsequent call to TryGetValue() to which you still need to pass exactly the same object. It will be easy to forget that. Here, both calls next to each other take the same Sample object, which, in my opinion, makes it easier to understand and maintain.